README.md
# Ruby wrapper for pdf2htmlEX

`pdf2htmlEX` converts PDF to HTML while retaining text, format and style as much as possible by making use of HTML5, JavaScript and modern CSS features. Even difficult content like PDFs with embedded fonts, multicolumn documents, scientific papers with complicated figures and mathematical formulas will mostly be represented correctly. Fallback mode generates HTML pages which do not require any JavaScript to view them correctly at the expense of a larger file size.

## Installation

You will need `pdf2htmlEX` installed. If you are using Mac OS X, I recommend installing pdf2htmlEX with Homebrew by running the following:

    $ brew install pdf2htmlex

If you are using Debian, you can install the pdf2htmlex package like so:

    $ apt install pdf2htmlex

If you've docker env setup, just install it via docker:

    $ alias pdf2htmlex="docker run -ti --rm -v ~/pdf:/tmp/pdf iapain/pdf2htmlex pdf2htmlEX"

After that just add this line to your application's Gemfile:

```ruby
gem 'pdf2htmlex'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install pdf2htmlex

## Usage

Configuration:

```ruby
require 'pdf2htmlex'

Pdf2htmlex.config.executable = '/usr/local/bin/pdf2htmlEX' # Default value: 'pdf2htmlex'
```

To convert files use the `Pdf2htmlex::convert` method with the PDF file as the first argument, optional output filename as the second argument and list of options. Example:

```ruby
2.7.0 (main):0 > output = Pdf2htmlex.convert('./example/demo1.pdf')
=> #<Pathname:/var/folders/rn/1xx0_xsd089fldtvtnq1tk7m0000gn/T/d20210409-96258-1pmrw5z/demo1.html>
2.7.0 (main):0 > output = Pdf2htmlex.convert('./example/demo1.pdf', 'sample.html')
=> #<Pathname:/var/folders/rn/1xx0_xsd089fldtvtnq1tk7m0000gn/T/d20210409-96258-14l7p6u/sample.html>
2.7.0 (main):0 > output = Pdf2htmlex.convert('./example/demo1.pdf', 'demo.html', dest_dir: '/tmp')
=> #<Pathname:/tmp/demo.html>
```

All options:
```bash
first_page: '--first-page',                           # first page to convert (default: 1)
last_page: '--last-page',                             # last page to convert (default: 2147483647)
zoom: '--zoom',                                       # zoom ratio
fit_width: '--fit-width',                             # fit width to <fp> pixels
fit_height: '--fit-height',                           # fit height to <fp> pixels
use_cropbox: '--use-cropbox',                         # use CropBox instead of MediaBox (default: 1)
hdpi: '--hdpi',                                       # horizontal resolution for graphics in DPI (default: 144)
vdpi: '--vdpi',                                       # vertical resolution for graphics in DPI (default: 144)
embed: '--embed',                                     # specify which elements should be embedded into output
embed_css: '--embed-css',                             # embed CSS files into output (default: 1)
embed_font: '--embed-font',                           # embed font files into output (default: 1)
embed_image: '--embed-image',                         # embed image files into output (default: 1)
embed_javascript: '--embed-javascript',               # embed JavaScript files into output (default: 1)
embed_outline: '--embed-outline',                     # embed outlines into output (default: 1)
split_pages: '--split-pages',                         # split pages into separate files (default: 0)
dest_dir: '--dest-dir',                               # specify destination directory (default: ".")
css_filename: '--css-filename',                       # filename of the generated css file (default: "")
page_filename: '--page-filename',                     # filename template for split pages  (default: "")
outline_filename: '--outline-filename',               # filename of the generated outline file (default: "")
process_nontext: '--process-nontext',                 # render graphics in addition to text (default: 1)
process_outline: '--process-outline',                 # show outline in HTML (default: 1)
process_annotation: '--process-annotation',           # show annotation in HTML (default: 0)
process_form: '--process-form',                       # include text fields and radio buttons (default: 0)
printing: '--printing',                               # enable printing support (default: 1)
fallback: '--fallback',                               # output in fallback mode (default: 0)
tmp_file_size_limit: '--tmp-file-size-limit',         # Maximum size (in KB) used by temporary files, -1 for no limit. (default: -1)
embed_external_font: '--embed-external-font',         # embed local match for external fonts (default: 1)
font_format: '--font-format',                         # suffix for embedded font files (ttf,otf,woff,svg) (default: "woff")
decompose_ligature: '--decompose-ligature',           # decompose ligatures, such as ﬁ -> fi (default: 0)
auto_hint: '--auto-hint',                             # use fontforge autohint on fonts without hints (default: 0)
external_hint_tool: '--external-hint-tool',           # external tool for hinting fonts (overrides --auto-hint) (default: "")
stretch_narrow_glyph: '--stretch-narrow-glyph',       # stretch narrow glyphs instead of padding them (default: 0)
squeeze_wide_glyph: '--squeeze-wide-glyph',           # shrink wide glyphs instead of truncating them (default: 1)
override_fstype: '--override-fstype',                 # clear the fstype bits in TTF/OTF fonts (default: 0)
process_type3: '--process-type3',                     # convert Type 3 fonts for web (experimental) (default: 0)
heps: '--heps',                                       # horizontal threshold for merging text, in pixels (default: 1)
veps: '--veps',                                       # vertical threshold for merging text, in pixels (default: 1)
space_threshold: '--space-threshold',                 # word break threshold (threshold * em) (default: 0.125)
font_size_multiplier: '--font-size-multiplier',       # a value greater than 1 increases the rendering accuracy (default: 4)
space_as_offset: '--space-as-offset',                 # treat space characters as offsets (default: 0)
tounicode: '--tounicode',                             # how to handle ToUnicode CMaps (0=auto, 1=force, -1=ignore) (default: 0)
optimize_text: '--optimize-text',                     # try to reduce the number of HTML elements used for text (default: 0)
correct_text_visibility: '--correct-text-visibility', # try to detect texts covered by other graphics and properly arrange them (default: 0)
bg_format: '--bg-format',                             # specify background image format (default: "png")
svg_node_count_limit: '--svg-node-count-limit',       # if node count in a svg background image exceeds this limit, fall back this page to bitmap background; negative value means no limit. (default: -1)
svg_embed_bitmap: '--svg-embed-bitmap',               # 1: embed bitmaps in svg background; 0: dump bitmaps to external files if possible. (default: 1)
owner_password: '--owner-password',                   # owner password (for encrypted files)
user_password: '--user-password',                     # user password (for encrypted files)
no_drm: '--no-drm',                                   # override document DRM settings (default: 0)
clean_tmp: '--clean-tmp',                             # remove temporary files after conversion (default: 1)
tmp_dir: '--tmp-dir',                                 # specify the location of temporary directory. (default: "/var/folders/rn/1xx0_xsd089fldtvtnq1tk7m0000gn/T/")
data_dir: '--data-dir',                               # specify data directory (default: "/usr/local/Cellar/pdf2htmlex/0.14.6_24/share/pdf2htmlEX")
debug: '--debug',                                     # print debugging information (default: 0)
proof: '--proof',                                     # texts are drawn on both text layer and background for proof. (default: 0)
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/marcosgz/pdf2htmlex.


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
## [0.1.1] - 2021-04-09

### Added

- Include documentation url the gemspec
- fix broken CHANGELOG link of gemspec

## [0.1.0] - 2021-04-09

### Added

- Basic project structure to make the `Pdf2htmlex.convert()` function work

[0.1.1]: https://github.com/marcosgz/pdf2htmlex/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/marcosgz/pdf2htmlex/releases/tag/v0.1.0


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))

require 'pdf2htmlex'

RSpec.configure do |config|
  config.mock_framework = :mocha
end


spec/pdf2htmlex/convert_spec.rb
require 'spec_helper'

RSpec.describe Pdf2htmlex do
  class MockProcessStatus
    attr_accessor :exitstatus

    def initialize(exitstatus)
      @exitstatus = exitstatus
    end
  end

  def expect_command_to_match(*parts_of_command)
    Open3.expects(:capture3).with do |*command, _opts|
      command_str = command.join(' ')
      parts_of_command.each do |part_of_command|
        case part_of_command
        when Regexp
          expect(command_str).to match part_of_command
        when String
          expect(command_str).to include part_of_command
        else
          raise "Can't match that"
        end
      end
    end.returns(@result)
  end

  def stub_output(output)
    @result[0] = output
  end

  def stub_error(error)
    @result[1] = error
  end

  def stub_exitstatus(exitstatus)
    @result[2] = MockProcessStatus.new(exitstatus)
  end

  before do
    @result = ['', '', MockProcessStatus.new(0)]
    Open3.stubs(:capture3).returns(@result)
  end

  it 'allows the use of a custom path to the executable' do
    Pdf2htmlex.config.executable = '/custom/path/to/pdf2htmlex'

    expect(Pdf2htmlex.config.executable).to match /^\/custom\/path\/to\/pdf2htmlex/

    Pdf2htmlex.config.executable = 'pdf2htmlex'
  end

  describe '.convert' do
    it 'generates a destination directory' do
      Pdf2htmlex.stubs('make_tempdir').returns('/var/rn/1x')
      expect_command_to_match('--dest-dir /var/rn/1x')

      described_class.convert('file.pdf')
    end

    it 'uses the output filename' do
      expect_command_to_match('./file.pdf output.html')

      described_class.convert('./file.pdf', 'output.html')
    end

    it 'overwrites the destination directory' do
      expect_command_to_match('--dest-dir /tmp/target')

      described_class.convert('file.pdf', dest_dir: '/tmp/target')
    end

    it 'converts true boolean values to 1' do
      expect_command_to_match('--embed-javascript 1')

      described_class.convert('file.pdf', embed_javascript: true)
    end

    it 'converts false boolean values to 0' do
      expect_command_to_match('--embed-javascript 0')

      described_class.convert('file.pdf', embed_javascript: false)
    end

    it 'resolves Proc before convert' do
      expect_command_to_match('--embed-javascript 0')

      described_class.convert('file.pdf', embed_javascript: -> { 1 == 0 })
    end

    it 'converts multiple options and remove those without values' do
      expect_command_to_match(
        '--embed-javascript 0',
        '--dest-dir /tmp/target',
        '--font-format wtf',
        '--clean-tmp 1',
        './file.pdf out.html',
      )

      described_class.convert('./file.pdf', 'out.html',
        embed_javascript: -> { 1 == 0 },
        dest_dir: '/tmp/target',
        font_format: 'wtf',
        tounicode: -1,
        clean_tmp: true,
      )
    end

    it 'raises any errors from pdf2htmlex' do
      stub_error('error message')
      stub_exitstatus(1)
      expect { described_class.convert('filename') }.to raise_error(/pdf2htmlEX failed with: error message/)
    end

    it 'does not raise if pdf2htmlex exit status is 0' do
      stub_error('error message')
      expect { described_class.convert('filename') }.not_to raise_error
    end
  end

  [
    [{ last_page: 10 }, ['--last-page 10']],
    [{ zoom: 1.5 }, ['--zoom 1.5']],
    [{ fit_width: 800.0 }, ['--fit-width 800.0']],
    [{ fit_height: 600.0 }, ['--fit-height 600.0']],
    [{ use_cropbox: false }, ['--use-cropbox 0']],
    [{ hdpi: 144 }, ['--hdpi 144']],
    [{ vdpi: 144 }, ['--vdpi 144']],
    [{ embed: true }, ['--embed 1']],
    [{ embed_css: true }, ['--embed-css 1']],
    [{ embed_font: true }, ['--embed-font 1']],
    [{ embed_image: true }, ['--embed-image 1']],
    [{ embed_javascript: true }, ['--embed-javascript 1']],
    [{ embed_outline: true }, ['--embed-outline 1']],
    [{ split_pages: true }, ['--split-pages 1']],
    [{ dest_dir: './' }, ['--dest-dir ./']],
    [{ css_filename: 'style.css' }, ['--css-filename style.css']],
    [{ page_filename: 'p.html' }, ['--page-filename p.html']],
    [{ outline_filename: 'test' }, ['--outline-filename test']],
    [{ process_nontext: true }, ['--process-nontext 1']],
    [{ process_outline: true }, ['--process-outline 1']],
    [{ process_annotation: false }, ['--process-annotation 0']],
    [{ process_form: false }, ['--process-form 0']],
    [{ printing: true }, ['--printing 1']],
    [{ fallback: false }, ['--fallback 0']],
    [{ tmp_file_size_limit: -1 }, ['--tmp-file-size-limit -1']],
    [{ embed_external_font: false }, ['--embed-external-font 0']],
    [{ font_format: 'woff' }, ['--font-format woff']],
    [{ decompose_ligature: true }, ['--decompose-ligature 1']],
    [{ auto_hint: true }, ['--auto-hint 1']],
    [{ external_hint_tool: 'test' }, ['--external-hint-tool test']],
    [{ stretch_narrow_glyph: true }, ['--stretch-narrow-glyph 1']],
    [{ squeeze_wide_glyph: true }, ['--squeeze-wide-glyph 1']],
    [{ override_fstype: true }, ['--override-fstype 1']],
    [{ process_type3: true }, ['--process-type3 1']],
    [{ heps: true }, ['--heps 1']],
    [{ veps: true }, ['--veps 1']],
    [{ space_threshold: 0.125 }, ['--space-threshold 0.125']],
    [{ font_size_multiplier: 4 }, ['--font-size-multiplier 4']],
    [{ space_as_offset: true }, ['--space-as-offset 1']],
    [{ tounicode: -1 }, ['--tounicode -1']],
    [{ optimize_text: true }, ['--optimize-text 1']],
    [{ correct_text_visibility: true }, ['--correct-text-visibility 1']],
    [{ bg_format: 'png' }, ['--bg-format png']],
    [{ svg_node_count_limit: -1 }, ['--svg-node-count-limit -1']],
    [{ svg_embed_bitmap: true }, ['--svg-embed-bitmap 1']],
    [{ owner_password: 'user' }, ['--owner-password user']],
    [{ user_password: 'pass' }, ['--user-password pass']],
    [{ no_drm: true }, ['--no-drm 1']],
    [{ clean_tmp: true }, ['--clean-tmp 1']],
    [{ tmp_dir: '/tmp' }, ['--tmp-dir /tmp']],
    [{ data_dir: '/tmp' }, ['--data-dir /tmp']],
    [{ debug: true }, ['--debug 1']],
    [{ proof: true }, ['--proof 1']],
  ].each do |keywords, expected_options|
    specify do
      expect_command_to_match(*expected_options)

      described_class.convert('file.pdf', **keywords)
    end
  end

end


lib/pdf2htmlex.rb
# frozen_string_literal: true

# @see https://www.rubydoc.info/stdlib/open3/Open3.popen3
require 'open3'
# @see https://www.rubydoc.info/stdlib/pathname/Pathname
require 'pathname'
# @see https://ruby-doc.com/stdlib/libdoc/tmpdir/rdoc/Dir.html
require 'tmpdir'

require_relative 'pdf2htmlex/version'
require_relative 'pdf2htmlex/config'

# Ruby wrapper for the pdf2htmlEX tool
#
# `pdf2htmlEX` converts PDF to HTML while retaining text, format and style as much as
# possible by making use of HTML5, JavaScript and modern CSS features. Even difficult
# content like PDFs with embedded fonts, multicolumn documents, scientific papers with
# complicated figures and mathematical formulas will mostly be represented correctly.
# Fallback mode generates HTML pages which do not require any JavaScript to view them
# correctly at the expense of a larger file size.
#
# @see https://github.com/coolwanglu/pdf2htmlEX
module Pdf2htmlex
  class Error < StandardError; end

  extend self

  # Convert input PDF file to HTML
  #
  # @param input_pdf [String, Pathname] The path of PDF file
  # @param html_filename [String, Pathname, NilClass] The output HTML filename
  # @param options [Hash] List of pdf2htmlex options
  # @option [Integer] :first_page (default: 1) First page to convert
  # @option [Integer] :last_page (default: 2147483647) Last page to convert
  # @option [Float] :zoom Zoom ratio
  # @option [Float] :fit_width Fit width to <fp> pixels
  # @option [Float] :fit_height Fit height to <fp> pixels
  # @option [Boolean, Integer] :use_cropbo (default: true) Use CropBox instead of MediaBox
  # @option [Float] :hdpi (default: 144) Horizontal resolution for graphics in DPI
  # @option [Float] :vdpi (default: 144) Vertical resolution for graphics in DPI
  # @option [String] :embed Specify which elements should be embedded into output
  # @option [Boolean, Integer] :embed_css (default: true) Embed CSS files into output
  # @option [Boolean, Integer] :embed_font (default: true) Embed font files into output
  # @option [Boolean, Integer] :embed_image (default: true) Embed image files into output
  # @option [Boolean, Integer] :embed_javascript (default: true) Embed JavaScript files into output
  # @option [Boolean, Integer] :embed_outline (default: true) Embed outlines into output
  # @option [Boolean, Integer] :split_pages (default: false) Split pages into separate files
  # @option [String] :dest_dir (default: ".") Specify destination directory
  # @option [String] :css_filename (default: "") Filename of the generated css file
  # @option [String] :page_filename (default: "") Filename template for split pages
  # @option [String] :outline_filename (default: "") Filename of the generated outline file
  # @option [Boolean, Integer] :process_nontext (default: true) Render graphics in addition to text
  # @option [Boolean, Integer] :process_outline (default: true) Show outline in HTML
  # @option [Boolean, Integer] :process_annotation (default: false) Show annotation in HTML
  # @option [Boolean, Integer] :process_form (default: false) Include text fields and radio buttons
  # @option [Boolean, Integer] :printing (default: true) Enable printing support
  # @option [Boolean, Integer] :fallback (default: false) Output in fallback mode
  # @option [Integer] :tmp_file_size_limit (default: -1) Maximum size (in KB) used by temporary files, -1 for no limit.
  # @option [Boolean, Integer] :embed_external_font (default: true) Embed local match for external fonts
  # @option [String] :font_format (default: "woff") Suffix for embedded font files (ttf,otf,woff,svg)
  # @option [Boolean, Integer] :decompose_ligature (default: false) Decompose ligatures, such as ﬁ -> fi
  # @option [Boolean, Integer] :auto_hint (default: false) Use fontforge autohint on fonts without hints
  # @option [String] :external_hint_tool (default: "") External tool for hinting fonts (overrides --auto-hint)
  # @option [Boolean, Integer] :stretch_narrow_glyph (default: false) Stretch narrow glyphs instead of padding them
  # @option [Boolean, Integer] :squeeze_wide_glyph (default: true) Shrink wide glyphs instead of truncating them
  # @option [Boolean, Integer] :override_fstype (default: false) Clear the fstype bits in TTF/OTF fonts
  # @option [Boolean, Integer] :process_type3 (default: false) Convert Type 3 fonts for web (experimental)
  # @option [Float] :heps (default: 1) Horizontal threshold for merging text, in pixels
  # @option [Float] :veps (default: 1) Vertical threshold for merging text, in pixels
  # @option [Float] :space_threshold(default: 0.125)  Word break threshold (threshold * em)
  # @option [Integer] :font_size_multiplier (default: 4) A value greater than 1 increases the rendering accuracy
  # @option [Boolean, Integer] :space_as_offset (default: false) Treat space characters as offsets
  # @option [Integer] :tounicode (default: 0) How to handle ToUnicode CMaps (0=auto, 1=force, -1=ignore)
  # @option [Boolean, Integer] :optimize_text (default: false) Try to reduce the number of HTML elements used for text
  # @option [Boolean, Integer] :correct_text_visibility (default: false) Try to detect texts covered by other graphics and properly arrange them
  # @option [String] :bg_format (default: "png") Specify background image format
  # @option [Integer] :svg_node_count_limit (default: -1) if node count in a svg background image exceeds this limit, fall back this page to bitmap background; negative value means no limit
  # @option [Boolean, Integer] :svg_embed_bitmap (default: true) True: embed bitmaps in svg background; False: dump bitmaps to external files if possible.
  # @option [String] :owner_password Owner password (for encrypted files)
  # @option [String] :user_password User password (for encrypted files)
  # @option [Boolean, Integer] :no_drm (default: false) Override document DRM settings
  # @option [Boolean, Integer] :clean_tmp (default: true) Remove temporary files after conversion
  # @option [String] :tmp_dir (default: "/var/folders/rn/1xx0_xsd089fldtvtnq1tk7m0000gn/T/") Specify the location of temporary directory
  # @option [String] :data_dir (default: "/usr/local/Cellar/pdf2htmlex/0.14.6_24/share/pdf2htmlEX") Specify data directory
  # @option [Boolean, Integer] :debug (default: false) Print debugging information
  # @option [Boolean, Integer] :proof (default: false) Exts are drawn on both text layer and background for proof.
  # @raise [Pdf2htmlex::Error] catch errors from pdf2htmlEX into Pdf2htmlex::Error exception
  # @return [Pathname] The generated HTML file
  def convert(input_pdf, html_filename = nil, **options)
    options[:dest_dir] ||= make_tempdir

    output = output_html(options[:dest_dir], html_filename || input_pdf)
    command = normalize_options(options)
    command << input_pdf
    command << output.basename.to_s

    run_command(command)

    output
  end

  private

  def make_tempdir
    Dir.mktmpdir
  end

  # Generate a unique HTML filename within destination directory
  # @param base_dir [String] The directory path
  # @param target_filename [String] the
  def output_html(base_dir, target_filename)
    basename = File.basename(target_filename).sub(%r{\.(pdf|htm|html)$}i, '')
    filename = File.join(base_dir, format('%<fname>s.html', fname: basename))
    n = 0
    while File.exist?(filename)
      n += 1
      filename = File.join(base_dir, format('%<fname>s.%<fnum>d.html', fname: basename, fnum: n))
    end
    Pathname.new(filename)
  end

  def normalize_options(opts)
    [].tap do |arr|
      opts.each do |k, v|
        next if v.nil?
        next unless CMD_OPTIONS.key?(k)

        arr << CMD_OPTIONS[k] << cast_value(v)
      end
    end
  end

  def cast_value(value)
    case value
    when Proc then cast_value(value.call)
    when TrueClass then '1'
    when FalseClass then '0'
    else
      value
    end
  end

  def run_command(command, input = nil)
    opts = { binmode: true, stdin_data: input }

    command.unshift config.executable

    output, error, status = Open3.capture3(*command, opts)

    if status.exitstatus != 0
      raise Error, "pdf2htmlEX failed with: #{error}\nCommand: #{command.join(' ')}"
    end

    output
  end
end


lib/pdf2htmlex/config.rb
# frizen_string_literal: true

module Pdf2htmlex
  CMD_OPTIONS = {
    first_page: '--first-page',                           # first page to convert (default: 1)
    last_page: '--last-page',                             # last page to convert (default: 2147483647)
    zoom: '--zoom',                                       # zoom ratio
    fit_width: '--fit-width',                             # fit width to <fp> pixels
    fit_height: '--fit-height',                           # fit height to <fp> pixels
    use_cropbox: '--use-cropbox',                         # use CropBox instead of MediaBox (default: 1)
    hdpi: '--hdpi',                                       # horizontal resolution for graphics in DPI (default: 144)
    vdpi: '--vdpi',                                       # vertical resolution for graphics in DPI (default: 144)
    embed: '--embed',                                     # specify which elements should be embedded into output
    embed_css: '--embed-css',                             # embed CSS files into output (default: 1)
    embed_font: '--embed-font',                           # embed font files into output (default: 1)
    embed_image: '--embed-image',                         # embed image files into output (default: 1)
    embed_javascript: '--embed-javascript',               # embed JavaScript files into output (default: 1)
    embed_outline: '--embed-outline',                     # embed outlines into output (default: 1)
    split_pages: '--split-pages',                         # split pages into separate files (default: 0)
    dest_dir: '--dest-dir',                               # specify destination directory (default: ".")
    css_filename: '--css-filename',                       # filename of the generated css file (default: "")
    page_filename: '--page-filename',                     # filename template for split pages  (default: "")
    outline_filename: '--outline-filename',               # filename of the generated outline file (default: "")
    process_nontext: '--process-nontext',                 # render graphics in addition to text (default: 1)
    process_outline: '--process-outline',                 # show outline in HTML (default: 1)
    process_annotation: '--process-annotation',           # show annotation in HTML (default: 0)
    process_form: '--process-form',                       # include text fields and radio buttons (default: 0)
    printing: '--printing',                               # enable printing support (default: 1)
    fallback: '--fallback',                               # output in fallback mode (default: 0)
    tmp_file_size_limit: '--tmp-file-size-limit',         # Maximum size (in KB) used by temporary files, -1 for no limit. (default: -1)
    embed_external_font: '--embed-external-font',         # embed local match for external fonts (default: 1)
    font_format: '--font-format',                         # suffix for embedded font files (ttf,otf,woff,svg) (default: "woff")
    decompose_ligature: '--decompose-ligature',           # decompose ligatures, such as ﬁ -> fi (default: 0)
    auto_hint: '--auto-hint',                             # use fontforge autohint on fonts without hints (default: 0)
    external_hint_tool: '--external-hint-tool',           # external tool for hinting fonts (overrides --auto-hint) (default: "")
    stretch_narrow_glyph: '--stretch-narrow-glyph',       # stretch narrow glyphs instead of padding them (default: 0)
    squeeze_wide_glyph: '--squeeze-wide-glyph',           # shrink wide glyphs instead of truncating them (default: 1)
    override_fstype: '--override-fstype',                 # clear the fstype bits in TTF/OTF fonts (default: 0)
    process_type3: '--process-type3',                     # convert Type 3 fonts for web (experimental) (default: 0)
    heps: '--heps',                                       # horizontal threshold for merging text, in pixels (default: 1)
    veps: '--veps',                                       # vertical threshold for merging text, in pixels (default: 1)
    space_threshold: '--space-threshold',                 # word break threshold (threshold * em) (default: 0.125)
    font_size_multiplier: '--font-size-multiplier',       # a value greater than 1 increases the rendering accuracy (default: 4)
    space_as_offset: '--space-as-offset',                 # treat space characters as offsets (default: 0)
    tounicode: '--tounicode',                             # how to handle ToUnicode CMaps (0=auto, 1=force, -1=ignore) (default: 0)
    optimize_text: '--optimize-text',                     # try to reduce the number of HTML elements used for text (default: 0)
    correct_text_visibility: '--correct-text-visibility', # try to detect texts covered by other graphics and properly arrange them (default: 0)
    bg_format: '--bg-format',                             # specify background image format (default: "png")
    svg_node_count_limit: '--svg-node-count-limit',       # if node count in a svg background image exceeds this limit, fall back this page to bitmap background; negative value means no limit. (default: -1)
    svg_embed_bitmap: '--svg-embed-bitmap',               # 1: embed bitmaps in svg background; 0: dump bitmaps to external files if possible. (default: 1)
    owner_password: '--owner-password',                   # owner password (for encrypted files)
    user_password: '--user-password',                     # user password (for encrypted files)
    no_drm: '--no-drm',                                   # override document DRM settings (default: 0)
    clean_tmp: '--clean-tmp',                             # remove temporary files after conversion (default: 1)
    tmp_dir: '--tmp-dir',                                 # specify the location of temporary directory. (default: "/var/folders/rn/1xx0_xsd089fldtvtnq1tk7m0000gn/T/")
    data_dir: '--data-dir',                               # specify data directory (default: "/usr/local/Cellar/pdf2htmlex/0.14.6_24/share/pdf2htmlEX")
    debug: '--debug',                                     # print debugging information (default: 0)
    proof: '--proof',                                     # texts are drawn on both text layer and background for proof. (default: 0)
  }

  Config = Struct.new(:executable)

  def config
    @config ||= Config.new('pdf2htmlex')
  end
end
